centrallix/man/centrallix.8
.TH CENTRALLIX 8 "Nov 17, 2017" "Version 0.9.1" "Centrallix Manual"
.SH NAME
centrallix - the Centrallix Application Platform server daemon
.SH SYNOPSIS
.B centrallix
[\-c \fI<configfile>\fP] [\-d] [\-h] [\-q] [\-V]
.SH DESCRIPTION
.PP
The Centrallix Application Platform is a highly flexible environment
for developing rich internet applications (RIA's).  It features both
data and application abstraction, a SQL engine for joining diverse
data sources, declarative (not scripted) development, reports,
components, and style templates.
.PP
Centrallix's application abstraction means that app developers do not
code to the actual application deployment vehicle (presently RIA's via
HTML/HTTP).  This should allow Centrallix apps to be deployed using
other methods in the future, without requiring changes to the apps
themselves.
.PP
Centrallix's data abstraction means that data can be spread across
diverse data sources and still joined together using Centrallix's
built-in SQL engine, simplifying data integration needs.  It also
means that applications can be agnostic of the database backend that
is in use.  Furthermore, Centrallix apps themselves are written as
structured data, which could potentially be stored in a variety of
formats.
.PP
The system also includes a report writer and formatter, which supports
output in a variety of formats including plain text, PostScript, PDF, CSV
files, PCL, and HTML.
.SH OPTIONS
.l
Command-line options for the centrallix server binary:
.TP 8
.B  \-c \fIconfig file\fP
Specify a configuration file to use.  The default is typically either
/etc/centrallix.conf or /usr/local/etc/centrallix.conf, depending on
how Centrallix was installed.
.TP 8
.B  \-d
Run as a background service: fork and run in the background after
starting up.
.TP 8
.B  \-h
Print command-line options (help).
.TP 8
.B  \-q
Initialize quietly (don't print copyright header and module information).
.TP 8
.B \-V
Print Centrallix version and copyright information.
.SH FILES
.TP 8
.B /etc/centrallix.conf
The main configuration file for Centrallix.  See
.BR centrallix.conf (5)
.TP 8
.B /etc/centrallix/rootnode
The file that roots the Centrallix "ObjectSystem".  This file is an
object, typically of type "system/uxfile", which establishes where
the ObjectSystem "lives".  A typical location is /var/centrallix/os
on managed installs.  It is possible for the ObjectSystem to be
rooted in something other than a directory in the local filesystem,
though that is not typically done at the time of writing of this
document.
.TP 8
.B /etc/centrallix/types.cfg
A hierarchical MIME types definition file for Centrallix.  This file
defines types and file extensions, as well as relationships between
types.
.TP 8
.B /etc/centrallix/useragent.cfg
A definition file for user agents (browsers) that Centrallix's HTML
application generation supports, including lists of capabilities in
each user agent.
.TP 8
.B /usr/lib/libCentrallix.so
A shared library from the centrallix-lib package which contains
foundational functionality for Centrallix, including handling of
strings, arrays, hash tables, string tokenizing, threading, the
security-oriented "quoting printf" module, and more.
.TP 8
.B /usr/lib/libStParse.so
A shared library permitting the parsing of Centrallix "structure
files", without support for full expressions.  This library is used
by some other programs, but not by the Centrallix server binary, as
it has its own copy of the "structure file" parser which has full
expression support.
.TP 8
.B /usr/lib/centrallix/ (directory)
The location of loadable Centrallix modules, typically supplying
additional (optional) data connectivity drivers (objectsystem drivers),
or additional network interfaces other than HTTP.
.TP 8
.B /usr/include/cxlib/ (directory)
The location of centrallix-lib development header files, if installed.
.TP 8
.B /var/centrallix/os/ (directory)
The typical location of the Centrallix objectsystem root (see rootnode,
above).  The ObjectSystem is analogous to a "web root" or "document
root" for web servers.
.SH EXAMPLES
Below is an example /etc/centrallix/rootnode file that is typical of
a Centrallix install:
.RS 4
.nf

$Version=2$
rootnode "system/uxfile"
    {
    // This should point to your ObjectSystem root directory.
    path = "/var/centrallix/os";
    }

.fi
.RE
Here is a very simple Centrallix application, just consisting of a
blank page with a title.  The file might typically be located at:
/var/centrallix/os/hello_world.app
.RS 4
.nf

$Version=2$
hello_world "widget/page"
    {
    // Page title, shows up in browser titlebar
    title = "Hello, World!";

    // Design height and width, needed so that Centrallix can scale
    // the application properly if the user's browser is a different
    // size
    width = 800;
    height = 600;
    }

.fi
.RE
.SH "KNOWN BUGS"
Below are a few known issues with the Centrallix platform.
.IP 1.
The policy-based security subsystem (CXSS) is, at present, mostly absent.  This
means that, among other things, there is insufficient control on what
"senses" a user can use to open an object.  For instance, without
the security subsystem, filesystem-level permissions are insufficient
to make a distinction between a user modifying a CSV file as a whole
and that user modifying a field in a record in the CSV file (the former
would typically require a higher security clearance, whereas the latter
ensures the user doesn't damage the structure of the CSV file).
.IP 2.
The multithreading model currently used by Centrallix is non-preemptive.  This
means that, depending on the flexibility of the underlying library,
calls to external services (such as a database server) may block the
entire Centrallix process, giving rise to the potential for application
level deadlocks or denial of service situations.
.IP 3.
We are aware of some memory/resource leaks in the Centrallix codebase.
.SH "REPORTING BUGS"
E-mail bug reports to:  centrallix-devel at lists.sourceforge.net

If initial confidentiality is desired, sensitive security issues can
be sent directly to the primary maintainer of Centrallix at:
Greg.Beeley at LightSys.org
.SH AUTHORS
Greg Beeley, Josiah Ritchie, and other Centrallix developers.
.SH COPYRIGHT
Copyright (c) 1998-2017 LightSys Technology Services, Inc.
.SH "SEE ALSO"
.BR test_obj (1),
.BR centrallix.conf (5)


centrallix/CHANGELOG
Centrallix ChangeLog
-------------------------------------------------------------------------------
0.9.1 --> 0.9.2	    xx-xxx-xxxx     MAJOR DEVELOPMENT RELEASE

    Changes:

	1.  HTTP server implements X-Frame-Options anti-clickjacking header.

-------------------------------------------------------------------------------
0.9.0 --> 0.9.1	    13-Sep-2010     MAJOR DEVELOPMENT RELEASE - CATCHUP

    In this release, we're "catching up" with a lot of changes that have been
    made but never formally released.

    Changes:

	1.  Driver for MySQL databases [Jon C]
	2.  SQL updates:  DELETE support, multi-statement queries, performance
	    improvements, INCLUSIVE option for SUBTREE, WILDCARD data sources,
	    OBJECT keyword for data sources, and two-level aggregates.
	3.  Many improvements to the HTML widget set, too many to list here,
	    but including:  placeholder 'New row' for tables when creating a
	    new record (object), app/session global scope, 'description'
	    feature on editbox, 'link' styling for labels.
	4.  New expression (sql) functions: truncate, constrain, dateadd,
	    radians, degrees, sin, cos, tan, asin, acos, atan, atan2, sqrt,
	    square.
	5.  Report writer updates:  new CSV file output, and cx__download_as
	    to specify a generated filename for report downloads.
	6.  Many many bug fixes and efficiency improvements.
	7.  Improvements in developer documentation [Seth B]
	8.  QueryTree objects now support object creation and deletion.

-------------------------------------------------------------------------------
0.7.5 --> 0.9.0	    28-Mar-2008     MAJOR DEVELOPMENT RELEASE

    Changes:

	1.  Added component widgets, providing for reusable code.  Components
	    can be dynamically loaded, and can be multiply instantiated.
	2.  Added parameterization to applications and components.
	3.  Many many bug fixes and efficiency improvements.
	4.  SQL "INSERT INTO ... SELECT" support
	5.  SQL "UPDATE ... SET" support
	6.  Added autolayout (hbox/vbox) widget
	7.  Full Mozilla/Firefox support
	8.  Readonly filepro driver
	9.  Declarative "rule" widget for implementing various data and
	    application behavior, including relationships between osrc widgets.
	10. Several system-supplied components, including a generic form
	    component for auto-constructing a form and table based on the
	    information available about a data source in the objectsystem.
	11. "Repeat" widget used for auto-generating multiple widgets or trees
	    of widgets in an application or component based on the results of
	    a SQL query.  [David K]
	12. Enhanced query-by-form functionality [David K]

-------------------------------------------------------------------------------
0.7.4 --> 0.7.5	    (unreleased)    MAJOR DEVELOPMENT RELEASE

    Changes:

	1.  Centrallix-lib: added shared memory malloc, smmalloc, module and
	    a corresponding test suite [Greg]
	2.  Changed the calling convention on the *_init javascript functions
	    for widgets to be parameter order and number independent via the
	    use of inline static objects [Nathaniel]
	3.  Added an autolayout module to automatically size an application
	    for the target browser window [Nathaniel]
	4.  Some efficiency improvements in the javascript code [Greg]
	5.  Added SysInfo objectsystem driver, similar to /proc filesystem in
	    Linux [Greg]
	6.  Added "SELECT ... FROM SUBTREE /path/name" support to allow the
	    retrieval of an entire subtree of objects with one select
	    statement.
	7.  Added templating mechanism for applications to allow the
	    application of a "style" to an application, and reducing a lot of
	    cut-and-paste repetition for widget properties.
	8.  Changed name of 'htmlwindow' to 'childwindow' to remove the
	    terminology dependence on the DHTML/HTTP app delivery mechanism.
	9.  Added a working dropdown/popup menu widget, complete with 
	    separators, submenus, and main menubars.
	10. Added an inter-widget interface mechanism to standardize the way
	    widgets interact.
	11. Fixed a few memory leaks.

-------------------------------------------------------------------------------
0.7.3 --> 0.7.4	    01-Mar-2005     MAJOR DEVELOPMENT RELEASE

    Once again, this is a major development release - including a lot of new
    features and a lot of bug fixes.

    Changes:

	1.  Moderate Mozilla 1.x support and IE6 support now implemented, as
	    well as some user agent detection logic, capabilities matrix, and
	    browser compatibility layer.
	2.  Image widget added.
	3.  Obscure/Reveal subsystem for automatically managing the issue of
	    window or tab closure while data is unsaved, etc.
	4.  Serialization of loading of data from server to browser provides
	    a great deal more robustness.
	5.  V3 prtmgmt subsystem is now operational.
	6.  Report writer revamped; support for far more complex report outputs
	    including graphics, borders, etc.  
	    NOTE: incompatible changes made in report spec format ( .rpt file).
	7.  Matt McGill added a POP3 driver.
	8.  New scrollbar widget.
	9.  Widgets such as datetime, dropdown, checkbox, etc., can all receive
	    keyboard focus and participate in the tabbing order.
	10. Time-setting controls added to datetime widget
	11. Datetime, dropdown, and checkbox now allow NULL values.
	12. Presentation hints:  delivery to client working; some support for
	    certain hints on the client (such as allowchars/badchars, defaults,
	    etc.).  Also added 'alwaysdef' style hint for resetting the default
	    value on all modifications.  Better support for presentation hints
	    in sybase driver.
	13. objInfo() mechanism for obtaining metadata about an object.
	14. "widget tree" module added (MJM) - allows for three-step generation
	    process for apps, which allows composite widgets and in the near
	    future, component widgets.  Delivery of the entire structural
	    widget tree to the client is now supported.

centrallix/centrallix.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifdef HAVE_LIBDL
#include <dlfcn.h>
#endif
#include "centrallix.h"
#include "cxlib/mtask.h"
#include "obj.h"
#include "cxlib/xstring.h"
#include "cxlib/xarray.h"
#include "cxlib/xhash.h"
#include "stparse.h"
#include "cxlib/mtlexer.h"
#include <signal.h>
#include "wgtr.h"
#include "iface.h"
#include "cxss/cxss.h"
#include <openssl/ssl.h>
#include <openssl/rand.h>
#include <openssl/err.h>

/************************************************************************/
/* Centrallix Application Server System 				*/
/* Centrallix Core       						*/
/* 									*/
/* Copyright (C) 2002 LightSys Technology Services, Inc.		*/
/* 									*/
/* This program is free software; you can redistribute it and/or modify	*/
/* it under the terms of the GNU General Public License as published by	*/
/* the Free Software Foundation; either version 2 of the License, or	*/
/* (at your option) any later version.					*/
/* 									*/
/* This program is distributed in the hope that it will be useful,	*/
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	*/
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	*/
/* GNU General Public License for more details.				*/
/* 									*/
/* You should have received a copy of the GNU General Public License	*/
/* along with this program; if not, write to the Free Software		*/
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  		*/
/* 02111-1307  USA							*/
/*									*/
/* A copy of the GNU General Public License has been included in this	*/
/* distribution in the file "COPYING".					*/
/* 									*/
/* Module: 	centrallix.c           					*/
/* Author:	Greg Beeley (GRB)					*/
/* Creation:	May 1, 2002      					*/
/* Description:	This module contains common startup code which is used	*/
/*		by both test_obj and lsmain.				*/
/************************************************************************/



/*** The following variables are used to access build/version/stability
 *** information so that other modules don't have to be rebuilt with
 *** every compile - just this file in particular.
 ***/
char* cx__version = PACKAGE_VERSION;
int cx__build = BUILD;
int cx__subbuild = SUBBUILD;
char* cx__stability = STABILITY;
char* cx__years = YEARS;



/*** Instantiate the globals from centrallix.h 
 ***/
CxGlobals_t CxGlobals;
int CxSupportedInterfaces[] = {2, 0};

int
cx_internal_LoadModules(char* type)
    {
#ifdef WITH_DYNAMIC_LOAD
    pStructInf modules_inf;
    pStructInf one_module;
    char* modtype;
    char* modpath;
    char* modenable;
    pCxModule moduledata;
    int i,j,found;
    int n_loaded = 0;

	/** Find the modules section in the configuration file.  It is not
	 ** an error if the section is not there - just no module loading
	 ** will be done.  But, we'll warn about it anyhow.
	 **/
	modules_inf = stLookup(CxGlobals.ParsedConfig, "modules");
	if (!modules_inf) 
	    {
	    mssError(1,"CX","Warning - no modules section in config file");
	    return 0;
	    }

	/** Go through the modules list, looking for appropriate ones **/
	for(i=0;i<modules_inf->nSubInf;i++)
	    {
	    one_module = modules_inf->SubInf[i];
	    modtype=NULL;
	    modpath=NULL;
	    stAttrValue(stLookup(one_module,"path"),NULL,&modpath,0);
	    stAttrValue(stLookup(one_module,"type"),NULL,&modtype,0);
	    if (modtype && strcmp(modtype,type)) continue;
	    if (!modpath || !modtype)
		{
		if (!CxGlobals.QuietInit) printf("mod: driver %s failed\n", one_module->Name);
		mssError(1,"CX","Module '%s' must have path and type settings; load failed",one_module->Name);
		continue;
		}

	    /** Don't load it if it is disabled. **/
	    modenable = NULL;
	    stAttrValue(stLookup(one_module,"enable_module"), NULL, &modenable, 0);
	    if (modenable && !strcasecmp(modenable, "no"))
		{
		if (!CxGlobals.QuietInit)
		    {
		    printf("mod: %s %s disabled\n", modtype, one_module->Name);
		    }
		continue;
		}

	    /** Setup a structure for this fella **/
	    moduledata = (pCxModule)nmMalloc(sizeof(CxModule));
	    if (!moduledata) continue;
	    memset(moduledata, 0, sizeof(CxModule));

	    /** Try to load the bugger. **/
	    moduledata->dlhandle = dlopen(modpath, RTLD_NOW);
	    if (!moduledata->dlhandle)
		{
		nmFree(moduledata, sizeof(CxModule));
		if (!CxGlobals.QuietInit) printf("mod: %s %s failed\n", modtype, one_module->Name);
		mssError(1,"CX","Module '%s' could not be loaded",modpath);
		mssError(1,"CX","ERROR: %s",dlerror());
		continue;
		}

	    /** Lookup some key symbols. **/
	    moduledata->InitFn = dlsym(moduledata->dlhandle, "moduleInitialize");
	    if (!moduledata->InitFn)
		{
		dlclose(moduledata->dlhandle);
		nmFree(moduledata, sizeof(CxModule));
		if (!CxGlobals.QuietInit) printf("mod: %s %s failed\n", modtype, one_module->Name);
		mssError(1,"CX","Module '%s' has no Init function declared: %s",modpath, dlerror());
		continue;
		}
	    moduledata->DeinitFn = dlsym(moduledata->dlhandle, "moduleDeInitialize");
	    moduledata->Prefix = dlsym(moduledata->dlhandle, "modulePrefix");
	    moduledata->Description = dlsym(moduledata->dlhandle, "moduleDescription");
	    moduledata->Version = dlsym(moduledata->dlhandle, "moduleVersion");
	    moduledata->InterfaceVer = dlsym(moduledata->dlhandle, "moduleInterface");
	    if (!moduledata->InterfaceVer)
		{
		dlclose(moduledata->dlhandle);
		nmFree(moduledata, sizeof(CxModule));
		if (!CxGlobals.QuietInit) printf("mod: %s %s failed\n", modtype, one_module->Name);
		mssError(1,"CX","Module '%s' has no Interface Version declared: %s",modpath,dlerror());
		continue;
		}

	    /** Is the interface supported? **/
	    for(found=j=0;CxSupportedInterfaces[j];j++)
		{
		if (*(moduledata->InterfaceVer) == CxSupportedInterfaces[j])
		    {
		    found = 1;
		    break;
		    }
		}
	    if (!found)
		{
		dlclose(moduledata->dlhandle);
		nmFree(moduledata, sizeof(CxModule));
		if (!CxGlobals.QuietInit) printf("mod: %s %s failed\n", modtype, one_module->Name);
		mssError(1,"CX","Module '%s' has unsupported interface version %d (do you need to recompile/reinstall modules?)",
		    modpath, *(moduledata->InterfaceVer));
		continue;
		}

	    /** Attempt to initialize the module **/
	    if (moduledata->InitFn() < 0)
		{
		dlclose(moduledata->dlhandle);
		nmFree(moduledata, sizeof(CxModule));
		mssError(0,"CX","Module '%s' load failed",modpath);
		if (!CxGlobals.QuietInit) printf("mod: %s %s failed\n", modtype, one_module->Name);
		continue;
		}
	    
	    /** Ok, add this one to the list of modules. **/
	    moduledata->Next = CxGlobals.ModuleList;
	    CxGlobals.ModuleList = moduledata;
	    if (!CxGlobals.QuietInit)
		{
		printf("mod: %s %s (%s, %d.%d.%d)\n",
		    modtype,
		    moduledata->Prefix?*(moduledata->Prefix):"unknown",
		    moduledata->Description?*(moduledata->Description):"-",
		    moduledata->Version?((*(moduledata->Version))>>24):0,
		    moduledata->Version?(((*(moduledata->Version))>>16)&0xFF):0,
		    moduledata->Version?((*(moduledata->Version))&0xFFFF):0);
		}
	    n_loaded++;
	    }

	if (!CxGlobals.QuietInit && n_loaded > 0) printf("\n");
#endif

    return 0;
    }


/** cxShutdownThread - thread created from the SIGINT handler, 
***   call all registered shutdown handlers then exits
**/
void cxShutdownThread(void *v)
    {
    int i;
    if(CxGlobals.Flags & CX_F_SHUTTINGDOWN)
	return;
    CxGlobals.Flags |= CX_F_SHUTTINGDOWN;
    mssError(0,"CX","Centrallix is shutting down");
    for(i=0;i<xaCount(&CxGlobals.ShutdownHandlers);i++)
	{
	ShutdownHandlerFunc handler = (ShutdownHandlerFunc) xaGetItem(&CxGlobals.ShutdownHandlers,i);
	if(handler)
	    handler();
	}
    exit(0);
    thExit();
    }

int cxAddShutdownHandler(ShutdownHandlerFunc handler)
    {
    return xaAddItem(&CxGlobals.ShutdownHandlers,handler);
    }


void
cxRemovePidFile()
    {
    char pidbuf[16];
    int pidbuf_id;
    int fd;
    int cnt;
	

	if (CxGlobals.PidFile[0] && (CxGlobals.Flags & CX_F_SERVICE))
	    {
	    /** Double-check to see if we own it.  Another centrallix
	     ** process might have overwritten it.  Leave the file alone
	     ** in that case.
	     **
	     ** There is technically a race condition here -- but the
	     ** consequences are minor and we're doing a courtesy. :)
	     **/
	    fd = open(CxGlobals.PidFile, O_RDONLY, 0600);
	    if (fd >= 0)
		{
		cnt = read(fd, pidbuf, sizeof(pidbuf));
		if (cnt > 1 && cnt < sizeof(pidbuf))
		    {
		    pidbuf[cnt] = '\0';
		    pidbuf_id = strtol(pidbuf, NULL, 10);

		    /** Unlink it if the pids match **/
		    if (pidbuf_id == getpid())
			unlink(CxGlobals.PidFile);
		    }
		close(fd);
		}
	    
	    CxGlobals.PidFile[0] = '\0';
	    }

    return;
    }


int
cxInitialize(void* v)
    {
    pFile cxconf;
    pStructInf mss_conf;
    char* authmethod;
    char* authmethodfile;
    char* logmethod;
    char* logprog;
    int log_all_errors;
    char* ptr;
    int n;
    int fd;
    time_t tm;
    int pid;
    char rbuf[16];

	xaInit(&CxGlobals.ShutdownHandlers,4);

	/** set up the interrupt handler so we can shutdown properly **/
	mtAddSignalHandler(SIGINT,cxShutdownThread);
	mtAddSignalHandler(SIGTERM,cxShutdownThread);

	/** Add a shutdown handler to delete the pid file **/
	if (CxGlobals.PidFile[0])
	    cxAddShutdownHandler(cxRemovePidFile);

#ifdef _SC_CLK_TCK
        CxGlobals.ClkTck = sysconf(_SC_CLK_TCK);
#else
        CxGlobals.ClkTck = CLK_TCK;
#endif

	/** Startup message **/
	if (!CxGlobals.QuietInit)
	    {
	    printf("\n");
	    printf("Centrallix/%s build #%4.4d-%d [%s]\n\n", cx__version, cx__build, cx__subbuild, cx__stability);
	    printf("Copyright (C) %s LightSys Technology Services, Inc.\n", cx__years);
	    printf("An open source community developed project.  Provided with\n");
	    printf("ABSOLUTELY NO WARRANTY.  See the file 'COPYING' for details.\n");
	    printf("\n");
	    }

	/** Load the configuration file **/
	cxconf = fdOpen(CxGlobals.ConfigFileName, O_RDONLY, 0600);
	if (!cxconf)
	    {
	    printf("centrallix: could not open config file '%s'\n", CxGlobals.ConfigFileName);
	    thExit();
	    }
	CxGlobals.ParsedConfig = stParseMsg(cxconf, 0);
	if (!CxGlobals.ParsedConfig)
	    {
	    printf("centrallix: error parsing config file '%s'\n", CxGlobals.ConfigFileName);
	    thExit();
	    }
	fdClose(cxconf, 0);

	/** This setting can be dangerous apart from the RBAC security subsystem.
	 ** We default to Enabled here, but this is turned off in the default config.
	 **/
	if (stAttrValue(stLookup(CxGlobals.ParsedConfig, "enable_send_credentials"), &n, NULL, 0) != 0 || n != 0)
	    CxGlobals.Flags |= CX_F_ENABLEREMOTEPW;

	/** Init the security subsystem.
	 **/
	cxssInitialize();

	/** Init the session handler.  We have to extract the config data for this 
	 ** module ourselves, because mtsession is in the centrallix-lib, and thus can't
	 ** use the new stparse module's routines.
	 **/
	mss_conf = stLookup(CxGlobals.ParsedConfig, "mtsession");
	if (stAttrValue(stLookup(mss_conf,"auth_method"),NULL,&authmethod,0) < 0) authmethod = "system";
	if (stAttrValue(stLookup(mss_conf,"altpasswd_file"),NULL,&authmethodfile,0) < 0) authmethodfile = "/usr/local/etc/cxpasswd";
	if (stAttrValue(stLookup(mss_conf,"log_method"),NULL,&logmethod,0) < 0) logmethod = "stdout";
	if (stAttrValue(stLookup(mss_conf,"log_progname"),NULL,&logprog,0) < 0) logprog = "centrallix";
	log_all_errors = 0;
	if (stAttrValue(stLookup(mss_conf,"log_all_errors"),NULL,&ptr,0) < 0 || !strcmp(ptr,"yes")) log_all_errors = 1;
	mssInitialize(authmethod, authmethodfile, logmethod, log_all_errors, logprog);

	/** Initialize the various parts **/
	expInitialize();			/* expression processor/compiler */
	if (objInitialize() < 0) return -1;	/* OSML */
	snInitialize();				/* Node structure file handler */

	/** Let newmalloc in on the names of some structures **/
	nmRegister(sizeof(XString),"XString");
	nmRegister(sizeof(XArray),"XArray");
	nmRegister(sizeof(XHashTable),"XHashTable");
	nmRegister(sizeof(XHashEntry),"XHashEntry");
	nmRegister(sizeof(StructInf),"StructInf");
	nmRegister(sizeof(LxSession),"LxSession");
	nmRegister(sizeof(MTObject),"MTObject");
	nmRegister(sizeof(File),"File");
	nmRegister(sizeof(EventReq),"EventReq");
	nmRegister(sizeof(Thread),"Thread");
	nmRegister(sizeof(WgtrNode), "WgtrNode");

	/** Init the openssl library **/
	SSL_library_init();
	SSL_load_error_strings();
	fd = open("/dev/urandom", O_RDONLY | O_NOCTTY);
	if (fd >= 0)
	    {
	    read(fd, rbuf, 16);
	    RAND_add(rbuf, 16, (double)16.0);
	    close(fd);
	    }
	pid = getpid();
	RAND_add(&pid, 4, (double)0.25);
	tm = time(NULL);
	RAND_add(&tm, 4, (double)0.125);

	/** Init the multiquery system and drivers **/
	mqInitialize();				/* MultiQuery system */
	mqtInitialize();			/* tablegen query module */
	mqpInitialize();			/* projection query module */
	mqjInitialize();			/* join query module */
	mqisInitialize();			/* insert-select query mod */
	mquInitialize();			/* update statement query mod */
	mqusInitialize();			/* upsert statement query mod */
	mqdInitialize();			/* delete statement query mod */
	mqobInitialize();			/* orderby module */

	/** Init the objectsystem drivers **/
	sysInitialize();			/* Sys info driver */
	uxdInitialize();			/* UNIX filesystem driver */
#if 0
	sybdInitialize();			/* Sybase CT-lib driver */
	mysdInitialize();			/* MySQL driver */
#endif

	stxInitialize();			/* Structure file driver */
	qytInitialize();			/* Query Tree driver */
	qypInitialize();			/* Query Pivot driver */
	qyInitialize();				/* stored query (aka view) driver */
	rptInitialize();			/* report writer driver */
	uxpInitialize();			/* UNIX printer access driver */
	datInitialize();			/* flat ascii datafile (CSV, etc) */
	/*berkInitialize();*/			/* Berkeley Database support */
	uxuInitialize();			/* UNIX users list driver */
	audInitialize();			/* Audio file player driver */
	lnkInitialize();			/* Symlink driver */
	jsonInitialize();			/* JSON data driver */

	/** Init the reporting content drivers **/
#if 0
	pclInitialize();			/* PCL report generator */
	htpInitialize();			/* HTML report generator */
	fxpInitialize();			/* Epson FX report generator */
	txtInitialize();			/* text only report gen */
#endif

	prtInitialize();
	prt_htmlfm_Initialize();
	prt_csvfm_Initialize();
	prt_strictfm_Initialize();
	prt_pclod_Initialize();
	prt_textod_Initialize();
	prt_psod_Initialize();
	prt_fxod_Initialize();

	/** Initialize the wgtr module **/
	wgtrInitialize();

	/** Initialize the Interface module **/
	ifcInitialize();

	/** Init the modules being used if dynamic loading is disabled **/
	
#if (USE_DBL == CX_STATIC)
	dblInitialize();
#endif

#ifndef WITH_DYNAMIC_LOAD

#ifdef USE_HTTP
	httpInitialize();
#endif

#ifdef USE_MIME
	mimeInitialize();
#endif

#ifdef USE_SYBASE
	sybdInitialize();
#endif

#ifdef USE_MYSQL
	mysdInitialize();			/* MySQL driver */
#endif

#ifdef USE_XML
	xmlInitialize();
#endif

#endif

	/** Load any osdriver and querydriver modules. **/
	cx_internal_LoadModules("osdriver");
	cx_internal_LoadModules("qydriver");

    return 0;
    }


#ifndef CX_NO_HTMLGEN
int
cxHtInit()
    {

	/** Init the html-generation subsystem **/
	htrInitialize();			/* HTML generator */
	htruleInitialize();			/* rule module */
	htpageInitialize();			/* DHTML page generator */
	htspaneInitialize();		/* scrollable pane module */
	httreeInitialize();			/* treeview module */
	hthtmlInitialize();			/* html pane module */
	htconnInitialize();			/* connector nonvisual module */
	htbtnInitialize();			/* generic button module */
	htibtnInitialize();			/* image button module */
	httbtnInitialize();			/* text button module */
	htmenuInitialize();			/* dropdown/popup menu module */
	htsetInitialize();			/* frameset module */
	htvblInitialize();			/* variable nonvisual module */
	httabInitialize();			/* tab control / tab page module */
	htpnInitialize();			/* pane module */
	httblInitialize();			/* tabular data module */
	htwinInitialize();			/* draggable window module */
	htcbInitialize();			/* checkbox module */
	htrbInitialize();			/* radiobutton module */
	htebInitialize();			/* editbox module */
	httmInitialize();			/* timer nonvisual module */
	htexInitialize();			/* method exec module */
	htspnrInitialize();			/* spinner box module*/
	htfsInitialize();			/* form status module */
	htddInitialize();			/* dropdown htdrv module */
	htdtInitialize();			/* datetime htdrv module */
	httxInitialize();			/* textarea htdrv module */
	htclInitialize();			/* clock htdrv module */
	htcaInitialize();			/* calendar module */
	htsbInitialize();			/* scrollbar module */
	htimgInitialize();			/* image widget */
	htfbInitialize();			/* form bar composite widget test */
	htocInitialize();			/* object canvas widget */
	htmapInitialize();			/* object canvas widget */
	htfuInitialize();			/* file upload widget */

	htformInitialize();			/* forms module */
	htosrcInitialize();			/* osrc module */
	htalrtInitialize();			/* alert module */
	htlblInitialize();			/* label module */
	httermInitialize();			/* terminal module */
	hthintInitialize();			/* pres. hints module */
	htparamInitialize();			/* parameter module */
	htalInitialize();			/* autolayout module */
	htrptInitialize();			/* repeat module */
	htmsInitialize();			/* multiscroll module */

	htcmpdInitialize();			/* component declaration */
	htcmpInitialize();			/* component instance */

	/** Load any htdriver modules **/
	cx_internal_LoadModules("htdriver");

    return 0;
    }
#endif

#ifndef CX_NO_NETWORK
int
cxNetworkInit()
    {

	/** Init the network listeners. **/
#ifdef NHT_ENABLE
	nhtInitialize();			/* HTTP network listener */
#endif
#ifdef CGI_ENABLE
	cgiInitialize();			/* CGI "listener" */
#endif

#ifndef WITH_DYNAMIC_LOAD

#ifdef USE_NETLDAP
	nldapInitialize();			/* LDAP network interface */
#endif

#endif

	bnetInitialize();			/* BDQS network listener */

	/** Load any network driver modules **/
	cx_internal_LoadModules("netdriver");

    return 0;
    }
#endif
